File: sort/java/BogoSort.java
import java.security.SecureRandom;

public class BogoSort {
    public static void sort(int[] array, int attempts) {
        int cur = 0;
        SecureRandom random = new SecureRandom();
        while (!isSorted(array) && (attempts <= 0 || cur < attempts)) {
            if (attempts > 0) {
                cur++;
            }
            shuffle(array, random);
        }
    }

    private static boolean isSorted(int[] array) {
        for (int i = 0; i < array.length - 1; i++) {
            if (array[i] > array[i + 1]) {
                return false;
            }
        }
        return true;
    }

    private static void shuffle(int[] array, SecureRandom random) {
        for (int i = 0; i < array.length; i++) {
            int r = random.nextInt(array.length);
            int tmp = array[r];
            array[r] = array[i];
            array[i] = tmp;
        }
    }
}

File: sort/c/bogosort.c
#include "algorithms.h"
#include <stdbool.h>
#include <stdlib.h>
#include <time.h>

static bool issorted(int *array, size_t length) {
  for (size_t i = 1; i < length; i++) {
    if (array[i] < array[i - 1]) {
      return false;
    }
  }
  return true;
}

static void shuffle(int *array, size_t length) {
  for (size_t i = 0; i < length; i++) {
    swap(&array[i], &array[rand() % (i + 1)]);
  }
}

void bogosort(int *array, size_t length, size_t attempts) {
  size_t cur = 0;
  srand(time(NULL));
  while (!issorted(array, length) && (attempts == 0 || cur < attempts)) {
    if (attempts > 0) {
      cur++;
    }
    shuffle(array, length);
  }
}
